hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java

```diff
@@ -445,47 +447,49 @@ public MultiDeleteResponse multiDelete(@PathParam("bucket") String bucketName,
 
     OzoneBucket bucket = getBucket(bucketName);
     MultiDeleteResponse result = new MultiDeleteResponse();
+    Map<String, ErrorInfo> undeletedKeyResultMap;
+    List<String> deleteKeys = new ArrayList<>();
+
     if (request.getObjects() != null) {
       for (DeleteObject keyToDelete : request.getObjects()) {
-        long startNanos = Time.monotonicNowNanos();
-        try {
-          bucket.deleteKey(keyToDelete.getKey());
-          getMetrics().updateDeleteKeySuccessStats(startNanos);
-
-          if (!request.isQuiet()) {
-            result.addDeleted(new DeletedObject(keyToDelete.getKey()));
-          }
-        } catch (OMException ex) {
-          if (isAccessDenied(ex)) {
-            getMetrics().updateDeleteKeyFailureStats(startNanos);
-            result.addError(
-                new Error(keyToDelete.getKey(), "PermissionDenied",
-                    ex.getMessage()));
-          } else if (ex.getResult() != ResultCodes.KEY_NOT_FOUND) {
-            getMetrics().updateDeleteKeyFailureStats(startNanos);
-            result.addError(
-                new Error(keyToDelete.getKey(), "InternalError",
-                    ex.getMessage()));
-          } else {
+        deleteKeys.add(keyToDelete.getKey());
+      }
+      long startNanos = Time.monotonicNowNanos();
+      try {
+        undeletedKeyResultMap = bucket.deleteKeys(deleteKeys, true);
+        for (DeleteObject d : request.getObjects()) {
+          if (!(undeletedKeyResultMap.containsKey(d.getKey())) ||
+              undeletedKeyResultMap.get(d.getKey()).getCode().equals(ResultCodes.KEY_NOT_FOUND.name())) {
+            // if the key is not found, it is assumed to be successfully deleted
+            deleteKeys.remove(d.getKey());
             if (!request.isQuiet()) {
-              result.addDeleted(new DeletedObject(keyToDelete.getKey()));
+              result.addDeleted(new DeletedObject(d.getKey()));
             }
-            getMetrics().updateDeleteKeySuccessStats(startNanos);
+          } else if (undeletedKeyResultMap.containsKey(d.getKey()) &&
+              !undeletedKeyResultMap.get(d.getKey()).getCode().equals(ResultCodes.KEY_NOT_FOUND.name())) {
+            // All errors other than KEY_NOT_FOUND are returned
+            ErrorInfo error = undeletedKeyResultMap.get(d.getKey());
+            result.addError(new Error(d.getKey(), error.getCode(), error.getMessage()));
           }
```

Comment:
We can simplify this by storing the result of map lookup.

Also, when calling `equals()`, put the one known to be not-`null` first (or use `Objects.equals()` if either of them can be `null`).

```java
          ErrorInfo error = undeletedKeyResultMap.get(d.getKey());
          boolean deleted = error == null
              // if the key is not found, it is assumed to be successfully deleted
              || ResultCodes.KEY_NOT_FOUND.name().equals(error.getCode());
          if (deleted) {
            deleteKeys.remove(d.getKey());
            if (!request.isQuiet()) {
              result.addDeleted(new DeletedObject(d.getKey()));
            }
          } else {
            result.addError(new Error(d.getKey(), error.getCode(), error.getMessage()));
          }
```

If you accept the suggestion to avoid filling `deleteKeys` initially, then  call `deleteKeys.add()` in the `else` block instead of `deleteKeys.remove()` in the `if` block.

Reply:
Thanks for the suggestions! I have simplified the checks. 

> If you accept the suggestion to avoid filling deleteKeys initially, then call deleteKeys.add() in the else block instead of deleteKeys.remove() in the if block.

Regarding this suggestion, as I have mentioned above, we need to construct the string list of keys to make the bucket.deleteKeys() call. I am not sure if there is a way to do this without adding all the keys to the list. Please feel free to give any suggestions!